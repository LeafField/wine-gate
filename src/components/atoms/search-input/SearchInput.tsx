import React, { ChangeEvent, forwardRef } from "react";
import { CloseButton, TextInput } from "@mantine/core";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faMagnifyingGlass } from "@fortawesome/free-solid-svg-icons";
import { useStore } from "../../../store";

type Props = {
  error: string | null;
};

const SearchInput = forwardRef<HTMLInputElement, Props>(({ error }, ref) => {
  const { searchValue, setSearchValue } = useStore();

  const changeHandler = (e: ChangeEvent<HTMLInputElement>) => {
    setSearchValue(e.target.value);
  };

  return (
    <TextInput
      ref={ref}
      className="w-full focus:border-blue"
      leftSection={<FontAwesomeIcon icon={faMagnifyingGlass} />}
      rightSection={
        searchValue.length > 0 && (
          <CloseButton
            onClick={() => setSearchValue("")}
            className={`relative z-10 cursor-pointer`}
          />
        )
      }
      rightSectionPointerEvents="all"
      value={searchValue}
      onChange={changeHandler}
      placeholder="検索"
      name="search"
      aria-label="search"
      error={error}
      withErrorStyles={false}
      id="search"
    />
  );
});

export default SearchInput;

SearchInput.displayName = "SearchInput";
