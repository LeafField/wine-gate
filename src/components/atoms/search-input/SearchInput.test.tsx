import { render, fireEvent, screen } from "@testing-library/react";
import SearchInput from "./SearchInput";
import { MantineProvider } from "@mantine/core";

describe("SearchInputの単体テスト", () => {
  beforeAll(() => {
    Object.defineProperty(window, "matchMedia", {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // deprecated
        removeListener: jest.fn(), // deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  });

  test("コンポーネントがクラッシュせずにレンダリングされる", () => {
    render(<SearchInput />, { wrapper: MantineProvider });
    const inputElement = screen.getByRole("textbox", {
      name: "search",
    });
    expect(inputElement).toBeInTheDocument();
  });

  test("ユーザーの入力が正しく反映される", () => {
    render(<SearchInput />, { wrapper: MantineProvider });
    const inputElement = screen.getByRole<HTMLInputElement>("textbox", {
      name: "search",
    });
    fireEvent.change(inputElement, { target: { value: "test" } });
    expect(inputElement.value).toBe("test");
  });

  test("クローズボタンをクリックすると入力がクリアされる", () => {
    render(<SearchInput />, { wrapper: MantineProvider });
    const inputElement = screen.getByRole<HTMLInputElement>("textbox", {
      name: "search",
    });
    fireEvent.change(inputElement, { target: { value: "test" } });
    const closeButton = screen.getByRole("button");
    fireEvent.click(closeButton);
    expect(inputElement.value).toBe("");
  });

  test("スナップショットテスト", () => {
    const { container } = render(<SearchInput />, { wrapper: MantineProvider });
    expect(container).toMatchSnapshot();
  });
});
