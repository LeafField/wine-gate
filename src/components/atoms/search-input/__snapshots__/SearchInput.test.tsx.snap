// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInputの単体テスト スナップショットテスト 1`] = `
<div>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <div
    class="panel-inner focus:border-blue m-6c018570 mantine-Input-wrapper"
    data-variant="default"
    data-with-left-section="true"
    data-with-right-section="true"
    style="--input-left-section-pointer-events: none; --input-right-section-pointer-events: all;"
  >
    <div
      class="m-82577fc2 mantine-Input-section"
      data-position="left"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-magnifying-glass "
        data-icon="magnifying-glass"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376c-34.4 25.2-76.8 40-122.7 40C93.1 416 0 322.9 0 208S93.1 0 208 0S416 93.1 416 208zM208 352a144 144 0 1 0 0-288 144 144 0 1 0 0 288z"
          fill="currentColor"
        />
      </svg>
    </div>
    <input
      aria-invalid="false"
      aria-label="search"
      class="m-8fb7ebe7 mantine-Input-input"
      data-variant="default"
      name="search"
      placeholder="検索"
      value=""
    />
    <div
      class="m-82577fc2 mantine-Input-section"
      data-position="right"
    >
      <button
        class="mantine-focus-auto mantine-active m-220c80f2 hidden relative z-10 cursor-pointer m-86a44da5 mantine-CloseButton-root m-87cf2631 mantine-UnstyledButton-root"
        data-variant="subtle"
        type="button"
      >
        <svg
          fill="none"
          viewBox="0 0 15 15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M11.7816 4.03157C12.0062 3.80702 12.0062 3.44295 11.7816 3.2184C11.5571 2.99385 11.193 2.99385 10.9685 3.2184L7.50005 6.68682L4.03164 3.2184C3.80708 2.99385 3.44301 2.99385 3.21846 3.2184C2.99391 3.44295 2.99391 3.80702 3.21846 4.03157L6.68688 7.49999L3.21846 10.9684C2.99391 11.193 2.99391 11.557 3.21846 11.7816C3.44301 12.0061 3.80708 12.0061 4.03164 11.7816L7.50005 8.31316L10.9685 11.7816C11.193 12.0061 11.5571 12.0061 11.7816 11.7816C12.0062 11.557 12.0062 11.193 11.7816 10.9684L8.31322 7.49999L11.7816 4.03157Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
